enforcer.go

```diff
@@ -419,78 +422,54 @@ func (e *Enforcer) enforce(matcher string, explains *[]string, rvals ...interfac
 				len(rvals),
 				rvals)
 		}
-		for i, pvals := range e.model["p"]["p"].Policy {
-			// log.LogPrint("Policy Rule: ", pvals)
-			if len(e.model["p"]["p"].Tokens) != len(pvals) {
-				return false, fmt.Errorf(
-					"invalid policy size: expected %d, got %d, pvals: %v",
-					len(e.model["p"]["p"].Tokens),
-					len(pvals),
-					pvals)
-			}
 
-			parameters.pVals = pvals
-
-			if hasEval {
-				ruleNames := util.GetEvalValue(expString)
-				var expWithRule = expString
-				for _, ruleName := range ruleNames {
-					if j, ok := parameters.pTokens[ruleName]; ok {
-						rule := util.EscapeAssertion(pvals[j])
-						expWithRule = util.ReplaceEval(expWithRule, rule)
-					} else {
-						return false, errors.New("please make sure rule exists in policy when using eval() in matcher")
+		if e.parallel {
+			ctx, cancel := context.WithCancel(context.Background())
+
+			var wg sync.WaitGroup
+			wg.Add(len(e.model["p"]["p"].Policy))
+			errs := make(chan error, len(e.model["p"]["p"].Policy))
+
+			for i, pvals := range e.model["p"]["p"].Policy {
+				go func(i int, pvals []string, parameters enforceParameters, expression *govaluate.EvaluableExpression) {
+					defer wg.Done()
+
+					select {
+					case <-ctx.Done():
+						return
+					default:
 					}
 
-					expression, err = govaluate.NewEvaluableExpressionWithFunctions(expWithRule, functions)
+					parameters.pVals = pvals
+					var err error
+					policyEffects[i], matcherResults[i], err = e.evaluatePolicy(parameters, expression, expString, functions)
 					if err != nil {
-						return false, fmt.Errorf("p.sub_rule should satisfy the syntax of matcher: %s", err)
+						errs <- err
+						cancel()
 					}
-				}
-
+				}(i, pvals, parameters, expression)
 			}
 
-			result, err := expression.Eval(parameters)
-			// log.LogPrint("Result: ", result)
+			wg.Wait()
 
-			if err != nil {
+			select {
+			case err := <-errs:
 				return false, err
```

Comment:
I think use the following code instead of your code is better.
```golang
close(errs)
err := <-errs
if err != nil {
    return false, err
}

Reply:
Okay, updated.